client/scala/src/main/scala/ai/verta/blobs/dataset/S3.scala

```diff
@@ -0,0 +1,168 @@
+package ai.verta.blobs.dataset
+
+import ai.verta.swagger._public.modeldb.versioning.model._
+
+import com.amazonaws.regions.Regions;
+import com.amazonaws.services.s3._
+import com.amazonaws.services.s3.model._
+
+import collection.JavaConverters._
+
+import scala.collection.mutable.HashMap
+import scala.util.{Failure, Success, Try}
+import scala.annotation.tailrec
+
+/** Captures metadata about S3 objects
+ *  Please set up AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, and AWS_REGION environment variables before use.
+ *  To create a new instance, use the constructor taking a list of S3 Locations
+ *  {{{
+ *  val s3Blob: Try[S3] = S3(List(S3Location("some-path-1").get, S3Location("some-path-2").get))
+ *  }}}
+ */
+case class S3(
+  private val metadataList: List[Tuple2[String, FileMetadata]],
+  private val versionList: List[Tuple2[String, String]]
+) extends Dataset {
+  protected var contents = HashMap(metadataList: _*)
+  private var versionMap = HashMap(versionList: _*)
+
+  /** Get the version id of a file
+   *  @param path: S3 URL of a file in the form "s3://<bucketName>/<key>"
+   *  @return the version id of the file
+   */
+  def getVersionId(path: String) = versionMap.get(path)
+
+  override def equals(other: Any) = other match {
+    case other: S3 => contents.equals(other.contents) && versionMap.equals(other.versionMap)
+    case _ => false
+  }
+}
+
+/** Companion object to handle interaction with versioning blob */
+object S3 {
+  private val s3: AmazonS3 = AmazonS3ClientBuilder.standard().build()
+
+  /** Constructor that user should use:
+   */
+  def apply(paths: List[S3Location]): Try[S3] = {
```

Comment:
1. Provide a constructor that receives a simple location too
2. Provide a reducer that combines two S3 locations into a new one (also do that for Path)

Reply:
Ok will do. Shouldn't be too hard.

Reply:
For 2 do you mean reducing two S3 blob instance into one right? Combining S3Locations themselves doesn't make sense to me. I'll work on that instead.

Reply:
Reducers implemented (via apply factory method) and tested.